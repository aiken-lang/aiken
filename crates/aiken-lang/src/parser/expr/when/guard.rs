use crate::{
    ast,
    parser::{definition, error::ParseError, token::Token},
};
use chumsky::prelude::*;

pub fn parser() -> impl Parser<Token, ast::UntypedClauseGuard, Error = ParseError> {
    recursive(|expression| {
        let var_parser = select! {
            Token::Name { name } => name,
            Token::UpName { name } => name,
        }
        .map_with_span(|_name, _span| ast::UntypedClauseGuard {});

        let constant_parser = definition::constant::value().map(|_| ast::UntypedClauseGuard {});

        let block_parser = expression
            .clone()
            .delimited_by(just(Token::LeftParen), just(Token::RightParen));

        let leaf_parser = choice((var_parser, constant_parser, block_parser)).boxed();

        let unary_op = just(Token::Bang);

        let unary = unary_op
            .map_with_span(|op, span| (op, span))
            .repeated()
            .then(leaf_parser)
            .foldr(|(_, _span), _value| ast::UntypedClauseGuard {})
            .boxed();

        let comparison_op = choice((
            just(Token::EqualEqual).to(ast::BinOp::Eq),
            just(Token::NotEqual).to(ast::BinOp::NotEq),
            just(Token::Less).to(ast::BinOp::LtInt),
            just(Token::Greater).to(ast::BinOp::GtInt),
            just(Token::LessEqual).to(ast::BinOp::LtEqInt),
            just(Token::GreaterEqual).to(ast::BinOp::GtEqInt),
        ));

        let comparison = unary
            .clone()
            .then(comparison_op.then(unary).repeated())
            .foldl(|_left, (_op, _right)| ast::UntypedClauseGuard {})
            .boxed();

        let and_op = just(Token::AmperAmper);
        let conjunction = comparison
            .clone()
            .then(and_op.then(comparison).repeated())
            .foldl(|_left, (_tok, _right)| ast::UntypedClauseGuard {});

        let or_op = just(Token::VbarVbar);
        conjunction
            .clone()
            .then(or_op.then(conjunction).repeated())
            .foldl(|_left, (_tok, _right)| ast::UntypedClauseGuard {})
    })
}
