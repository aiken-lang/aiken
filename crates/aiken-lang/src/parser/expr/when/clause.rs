use chumsky::prelude::*;
use vec1::vec1;

use crate::{
    ast,
    expr::UntypedExpr,
    parser::{error::ParseError, pattern, token::Token},
};

use super::guard;

pub fn parser(
    expression: Recursive<'_, Token, UntypedExpr, ParseError>,
) -> impl Parser<Token, ast::UntypedClause, Error = ParseError> + '_ {
    pattern()
        .then(just(Token::Vbar).ignore_then(pattern()).repeated().or_not())
        .then(choice((
            just(Token::If)
                .ignore_then(guard())
                .or_not()
                .then_ignore(just(Token::RArrow)),
            just(Token::If)
                .ignore_then(take_until(just(Token::RArrow)))
                .validate(|_value, span, emit| {
                    emit(ParseError::invalid_when_clause_guard(span));
                    None
                }),
        )))
        // TODO: add hint "Did you mean to wrap a multi line clause in curly braces?"
        .then(expression)
        .map_with_span(
            |(((pattern, alternative_patterns_opt), guard), then), span| {
                let mut patterns = vec1![pattern];
                patterns.append(&mut alternative_patterns_opt.unwrap_or_default());
                ast::UntypedClause {
                    location: span,
                    patterns,
                    guard,
                    then,
                }
            },
        )
}

#[cfg(test)]
mod tests {
    use crate::assert_expr;

    #[test]
    fn when_clause_todo() {
        assert_expr!(
            r#"
            when val is {
              Bar1{..} -> True
              Bar2{..} -> todo @"unimplemented"
            }
            "#
        );
    }

    #[test]
    fn when_clause_solo_error() {
        assert_expr!(
            r#"
            when val is {
              Bar1{..} -> fail
            }
            "#
        );
    }

    #[test]
    fn when_clause_double_todo() {
        assert_expr!(
            r#"
            when val is {
              Bar1{..} -> todo
              Bar2{..} -> todo
            }
            "#
        );
    }
}
