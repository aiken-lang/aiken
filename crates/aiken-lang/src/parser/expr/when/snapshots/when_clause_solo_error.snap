---
source: crates/aiken-lang/src/parser/expr/when/clause.rs
description: "Code:\n\nwhen val is {\n  Bar1{..} -> fail\n}\n"
---
When {
    location: 0..34,
    subject: Var {
        location: 5..8,
        name: "val",
    },
    clauses: [
        UntypedClause {
            location: 16..32,
            patterns: [
                Constructor {
                    is_record: true,
                    location: 16..24,
                    name: "Bar1",
                    arguments: [],
                    module: None,
                    constructor: (),
                    with_spread: true,
                    tipo: (),
                },
            ],
            guard: None,
            then: Trace {
                kind: Error,
                location: 28..32,
                then: ErrorTerm {
                    location: 28..32,
                },
                text: String {
                    location: 28..32,
                    value: "aiken::error",
                },
            },
        },
    ],
}
